### PR TITLE
WIP TorchFX

### PR DESCRIPTION
**Pull request by ramiro050**
Toy implementation of a translator from the `torch.fx` module graph representation into `MLIR`.

## Current features

Currently I am able to take a module in the `torch.fx.GraphModule` representation and convert it into valid `MLIR` using the `torch` dialect. This implementation currently only supports the operations `torch.add`, `torch.sin`, and `torch.sigmoid`, but extending it to handle more operations should not be too difficult.

An example of the module translation can be found in `examples/example_add_sigmoid_sin.py`, which currently takes the `torch.fx.GraphModule`

```
graph():
    %x : [#users=1] = placeholder[target=x]
    %y : [#users=1] = placeholder[target=y]
    %sin : [#users=1] = call_function[target=torch.sin](args = (%x,), kwargs = {})
    %sigmoid : [#users=1] = call_function[target=torch.sigmoid](args = (%sin,), kwargs = {})
    %sigmoid_1 : [#users=1] = call_function[target=torch.sigmoid](args = (%y,), kwargs = {})
    %add : [#users=1] = call_function[target=torch.add](args = (%sigmoid, %sigmoid_1), kwargs = {})
    return add
```

and turns it into

```
module  {
  func @forward(%arg0: !torch.tensor, %arg1: !torch.tensor) -> !torch.tensor {
    %0 = torch.aten.sin %arg0 : !torch.tensor -> !torch.tensor
    %1 = torch.aten.sigmoid %0 : !torch.tensor -> !torch.tensor
    %2 = torch.aten.sigmoid %arg1 : !torch.tensor -> !torch.tensor
    %int1 = torch.constant.int 1
    %3 = torch.aten.add.Tensor %1, %2, %int1 : !torch.tensor, !torch.tensor, !torch.int -> !torch.tensor
    return %3 : !torch.tensor
  }
}
```

## Current Limitations

- I currently assume that all inputs to the `forward` method are `torch.tensor`s
- No source code location is being fed into the `MLIR` representation
- I assume a very flat structure for the `forward` method. In other words, there is no handling at the moment of things like `if` statements.
- I am currently using the `torch_mlir.ModuleBuilder` constructor simply because it is the only way I know of registering the `torch` dialect from `python`. It would be good to know if there is a `python` binding for registering dialects.
- The generated `forward` method is turned into a function because I am not yet handling the `MLIR` code generation for the `torch.nn.Module` where the `forward` method is expected to reside.

## Future steps

- Add documentation
- Implement a class annotator similar to the one currently in `mlir-npcomp` (https://github.com/llvm/mlir-npcomp/blob/main/frontends/pytorch/csrc/builder/class_annotator.cpp) to annotate the `torch.fx.GraphModule` representation
- Add generation of `MLIR` code that represents the `torch.nn.Module`, for this is required by the `npcomp.compiler.pytorch.backend.frontend_lowering`. At the moment only the `forward` method is generated in `MLIR`

@silvasean and @stellaraccident, would love any feedback, especially regarding the future steps. Am I headed in a good direction?

### CODE REVIEW

**Review by silvasean — CHANGES_REQUESTED** [review]
Thanks, this looks like a great start! You are very fast :)

> 
> ## Current Limitations
> * I currently assume that all inputs to the `forward` method are `torch.tensor`s

Make to remove that assumption before submitting (failing compilation with a nice diagnostic saying "type Foo is not supported" is fine).

> * No source code location is being fed into the `MLIR` representation

This is one of those things that's worth digging into from the beginning, since it's really hard to add back afterwards in my experience.

> * I assume a very flat structure for the `forward` method. In other words, there is no handling at the moment of things like `if` statements.

Generally speaking in the compiler, never assume anything unless it is a precondition enforced by the MLIR verifier (or the TorchFX verifier, if they have one). Failing with a clear diagnostic is fine though (getting the diagnostic / error-handling patterns set out nicely in the first patch is a good idea -- usually bad diagnostics (== bad user experience) are just due to it not being easy/consistent to add a new diagnostic).

> * I am currently using the `torch_mlir.ModuleBuilder` constructor simply because it is the only way I know of registering the `torch` dialect from `python`. It would be good to know if there is a `python` binding for registering dialects.

@stellaraccident is there any other way to do this from Python-land? If not we can just expose a standalone function.
Regardless, this seems fine for now (do annotate it with a TODO though to clean it up)

> * The generated `forward` method is turned into a function because I am not yet handling the `MLIR` code generation for the `torch.nn.Module` where the `forward` method is expected to reside.

My experience with the ivalue_importer is that it's worth getting the module structure imported first, and then layering the function import into it second. Otherwise you aren't really capturing a translation unit that is semantically meaningful in the source language (the semantics of a torch.fx.GraphModule is a whole module, not just a single function). For context, I had to go back and entirely rewrite what is now node_importer (analogous to what you have in the current patch) after implementing the ivalue_importer because it ended up making a lot of assumptions / wasn't layered right for what was actually needed.

**Review by stellaraccident — COMMENTED** [review]
This is really a great start!

Regarding dialect registration, that is annoying and our conventions for it emerged after this project was set up.

Here's an example of the pattern I finally settled on:

Python side: https://github.com/google/iree/blob/475045245a9ac0f913e368bff671c86cb9d36af9/llvm-external-projects/iree-dialects/python/mlir/dialects/iree.py#L8
Pybind side: https://github.com/google/iree/blob/475045245a9ac0f913e368bff671c86cb9d36af9/llvm-external-projects/iree-dialects/python/IREEDialectsModule.cpp#L49
CAPI side (should already be there): https://github.com/google/iree/blob/475045245a9ac0f913e368bff671c86cb9d36af9/llvm-external-projects/iree-dialects/include/iree-dialects-c/Dialects.h#L21

Basically, make the ODS generated module for a dialect import all symbols from a same named, native module, which includes the function `register_dialect` which does the deed. It uses dialect accessors from the CAPI, which should already be there, but since this project was one of the first done, some of the newer refinements haven't been backported.

File: frontends/pytorch/examples/torchfx2iree/builder.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        assert f_node.op == 'call_function'
+        args = [self.env[arg_node] for arg_node in f_node.args]
+
+        if f_node.target == torch.sigmoid:
+            assert len(args) == 1, "`sigmoid` only takes one argument"
+            result = torch_d.AtenSigmoidOp(self.get_torch_tensor_type(),
+                                           args[0],
+                                           loc=self.loc,
+                                           ip=ip).result
+        elif f_node.target == torch.sin:
```
Comment (stellaraccident) [inline frontends/pytorch/examples/torchfx2iree/builder.py:64]: If you end up keeping this big if/else chain, probably better to hoist it to a lookup table that resolves to a callback of some kind. Various ways to structure that. For extensibility, I somewhat prefer the decorator approach:

```
TARGET_BUILDERS = dict()

def target_builder(f=None, *, target):
  if not f: return functools.partial(target=target)
  def wrapped(...):
    TARGET_BUILDERS[target] = f
    return f

@target_builder(target=aten.sin)
def aten_sin_builder(...)
  ...
```

May need fiddling but something like that.

**Comment by powderluv** [comment]
We probably want to follow the acc_tracer  (https://github.com/pytorch/pytorch/tree/master/torch/fx/experimental/fx_acc) closely so we can get the benefits of the operator normalization for free

**Comment by silvasean** [comment]
> My experience with the ivalue_importer is that it's worth getting the module structure imported first, and then layering the function import into it second. Otherwise you aren't really capturing a translation unit that is semantically meaningful in the source language (the semantics of a torch.fx.GraphModule is a whole module, not just a single function). For context, I had to go back and entirely rewrite what is now node_importer (analogous to what you have in the current patch) after implementing the ivalue_importer because it ended up making a lot of assumptions / wasn't layered right for what was actually needed.

I looked more closely. The semantics of a torch.fx.GraphModule is literally just a single graph/function, and not a whole torch.nn.Module surprisingly. So I think just importing the graph into a function is fine.

**Comment by ramiro050** [comment]
## PR Update

### Comments that have been incorporated

- [x] I no longer assume that the input types for forward are always tensors. Using `annotator.py`, the user can annotate `forward` with a variety of types.
- [x] The flat structure assumption for `forward` is actually the way `torch.fx` works. There is no conditional branching in the `torch.fx.Graph` representation.
- [x] The builder now takes the class structure into account. Although, the comment Sean just made is making me rethink this entire design. I think I will redesign the function builder to be like the class builder (initialize the variables you need as you encounter them), and then get rid of the class builder entirely. This should simplify the code significantly.
- [x] No meta-programming for `torch.functions->mlir` yet, but I have used the decorator method Stella mentioned.

### Comments that have _not_ been incorporated

- [ ] I have a WIP implementation for getting source code information, but it is in very early stages. There are a few challenges regarding extraction of source code information, the main one being that there is no guarantee that the graph inputted by the user has not been modified to represent a completely different code.
- [ ] I am still using the `ModuleBuilder` to register the dialects. I'll look into the method Stella mentioned for getting a python binding.

## New things

### annotator.py

A simple design for an annotator that can work in an analogous form to the `class_annotator` used for `torch.jit.script`s. This annotator takes advantage of the fact that one can add keyword arguments to each node of the `torch.fx.Graph`. Therefore, the graph itself holds all the type information before being sent to the `MLIR` builder.

### builder.py

The work of building the MLIR code is now split into two main sections. First we have the class builder, which is in charge of generating the MLIR code that represents the `torch.nn.Module` class, with declarations for each attribute and method. The builder is also in charge of initializing the attributes of the class. This is done without using the `torch.fx.Graph` information, since all the information needed is in the `torch.nn.Module` itself.

The second part is the function builder which goes over the `torch.fx.Graph` and translates each instruction into `MLIR` in a very similar manner to how it was being done in the first draft.

The new changes I plan on taking, now that I know that the class structure is not necessary, is to fuse these two sections together. In other words, the function builder will translate each instruction into `MLIR`, and as attributes from the `torch.nn.Module` are needed, they will be initialized and passed into the function.

### loc.py

Initial implementation of a possible way of getting source information. The approach is to use Python's builtin functionality for generating an AST out of a python object, and then extracting information from the tree. The main limitation at the moment is that it might be difficult to fully match every instruction of the `torch.fx.Graph` with the AST representation, especially given the fact that the `acc` tracer makes changes to the graph, and the user can also do some rewriting. Therefore, there might be a limit to how much we can really know about the source information. I would love any advice or ideas for how we might move this forward.

### torch_mlir_types.py

A simple set of classes that help generate `MLIR` code for the types in the `torch` dialect. There are still many more types that need to be added.

## Future steps

- I have plenty of `TODO`s sprinkled all over the code that I need to take care of. However, the main thing that I plan on changing is the design of the builder to be centered around the `forward` function rather than the `torch.nn.Module` class. I think the experience I obtained making the class builder will make this a straight forward change that should take a few days.
- Begin incorporating source code location information where I can.
- Fix the earth moving bug. More on this below.


## Bug after earth moving

After rebasing to TOM, I am no longer able to run the example script `example_add_sigmoid_sin.py`. I get the following error

```shell
File "/nodclouddata/ramiro/mlir-npcomp/external/torch-mlir/examples/torchfx2iree/builder.py", line 270, in build_module
    return _ForwardFunctionBuilder(py_module, mlir_module).to_mlir()
  File "/nodclouddata/ramiro/mlir-npcomp/external/torch-mlir/examples/torchfx2iree/builder.py", line 173, in to_mlir
    func_op = builtin.FuncOp(f'{self.class_type_name}.forward',
  File "/nodclouddata/ramiro/mlir-npcomp/build/python_packages/npcomp_core/npcomp/dialects/_builtin_ops_ext.py", line 51, in __init__
    sym_name = StringAttr.get(str(name))
RuntimeError: An MLIR function requires a Context but none was provided in the call or from the surrounding environment. Either pass to the function with a 'context=' argument or establish a default using 'with Context():'
```

I am passing a context, but for some reason it is not being detected. I think this might be related to the fact that the torch dialect python bindings are no longer under the same directory as all the other dialects. Hopefully someone here knows what is going on.

**Comment by stellaraccident** [comment]
The upstream FuncOp constructor has a bug requiring a with-bound context. I
haven't tracked it down. Suggest surrounding that with `with context:` for
now. Upstream patch welcome too :)

**Comment by ramiro050** [comment]
The way I phrased the bug was a bit misleading. I am using a `with context:` before calling `FuncOp`. I had noticed also that passing `context` as an argument for `FuncOp` specifically does not work. The problem is that now `with context:` does not work either. I'll keep experimenting to see if I can find what's going wrong

**Comment by stellaraccident** [comment]
That is weird. I suspect something about conflicting versions of the python code/extensions. With all of the basicpy stuff gone, there aren't really any tests in tree that make sure the python environment is fully functional -- except what you are writing now.

Let me boot it up in a bit and see what I can find.

**Comment by ramiro050** [comment]
Here is a simple example of the bug.

```python
from npcomp import ir
# from mlir import ir # Uncomment this line for error
from npcomp.dialects import builtin

with ir.Context():
    bool_type = ir.Type.parse('f32')
    func_type = ([bool_type], [bool_type])
    func_op = builtin.FuncOp('forward', func_type,
                             loc=ir.Location.unknown())
```

 It seems that if I use the context from `mlir.ir`, `FuncOp` doesn't work. Only the context from `npcomp.ir` makes `FuncOp` work. The reason I use the `mlir.ir` one in my implementation is because I am using the context from `torch_mlir.ModuleBuilder` that has all the dialects registered.

**Comment by stellaraccident** [comment]
Ah, that is the problem: these are completely standalone implementations of
MLIR. While dangerous to mix like this in general, for this specific case,
it isn't working because reach has their own version of the context thread
local stack.

Really, there should only be one in the project (probably torch_mlir).
Typically, we organize the code so that relative imports get you what you
need and you never do a global import. This let's the project be relocated
and embedded into other compilers.

There's no workaround for this except to set it up right. Let me see if I
can untangle it and get you moving again. Likely just a couple of misplaced
bits.

**Comment by stellaraccident** [comment]
Ok, #310 should fix this for you. There are some other things that I noticed that we should normalize about the namespace before too much longer, but I wanted to err on the side of getting you a minimal fix that works.

To test, I did:
```
export PYTHONPATH=$PWD/python_packages/torch_mlir
python ~/scratch/test_torch_mlir.py
```

test_torch_mlir.py:
```
# from npcomp import ir
# Uncomment the next line for error
from mlir import ir
from mlir.dialects import builtin
from mlir.dialects import torch

with ir.Context() as ctx:
    torch.register_torch_dialect(ctx)
    float_type = ir.Type.parse('f32')
    func_type = ([float_type], [float_type])
    func_op = builtin.FuncOp('forward', func_type,
                             loc=ir.Location.unknown())

print("all good")
```

**Comment by stellaraccident** [comment]
I also filed this issue to track overall python namespace organization: https://github.com/llvm/mlir-npcomp/issues/311

**Comment by ramiro050** [comment]
## PR Update

This update mainly fixes the issues with the `ir.Context()` previously mentioned. The example `examples/example_add_tanh_sigmoid.py` now runs without a problem.

### Question

- What is the expected way of compiling the `torch MLIR` and running it for testing purposes? The `PassManager` from `mlir.passmanager` does not support the `torchscript-to-npcomp-backend-pipeline`, so I am not sure how to lower the `MLIR` as I was doing previously. What I have resorted to do in the meantime is to call a subprocess that runs `npcomp-opt` on the `MLIR` module, and then I compile using the `RefjitNpcompBackend`.  (I cannot use the `npcomp.passmanager` because that results in the same set of issues with `ir.Context()`)

### Comments that have been incorporated

- [x] I am no longer using `ModuleBuilder` to register the `torch` dialect. Not sure if this was always there, but it turns out there is a function called `register_torch_dialect` in `mlir.dialects.torch`.

**Comment by silvasean** [comment]
> What is the expected way of compiling the torch MLIR and running it for testing purposes?

There isn't currently a simple answer to that. Shelling out to npcomp-opt is a reasonably good solution for now. Keep in mind that strictly speaking any code that runs e2e is part of npcomp (not torch-mlir) because it depends on refbackend. I.e. code in torch-mlir cannot `import npcomp` or invoke `npcomp-opt`. I will be working on this soon to get a better story for "e2e examples".

For now, if you just put it in npcomp/examples, then this is fine to land (it just cannot live in torch-mlir if it depends on refbackend). We can work through the tougher issues related to locations, error handling, etc. later.

**Comment by silvasean** [comment]
You can do something like this (note the `str` -- it's important)
```
with npcomp.ir.Context() as ctx:
    npcomp.register_all_dialects(ctx)
    module = npcomp.ir.Module.parse(str(mb.module))

pm = PassManager.parse("torchscript-to-npcomp-backend-pipeline")
pm.run(module)
```

Current source of truth (e2e test framewo): https://github.com/llvm/mlir-npcomp/blob/ecc334123ca745497f422952f0041abbca1b7ec4/python/torch_support/npcomp_torchscript_e2e_test_configs/npcomp_backend.py#L92

**Comment by ramiro050** [comment]
> For now, if you just put it in npcomp/examples, then this is fine to land (it just cannot live in torch-mlir if it depends on refbackend). We can work through the tougher issues related to locations, error handling, etc. later.

Should I move just the `examples/example_add_tanh_sigmoid.py` file to `npcomp/examples`, or all the changes I have in this PR?

**Comment by silvasean** [comment]
at the present moment, layering-wise, I would recommend moving the entire PR into npcomp/examples/

**Comment by ramiro050** [comment]
## PR Update

- Added some documentation to `builder.py`
- The example script no longer makes a `subprocess` call to `npcomp-opt`

I still need to test my code after rebasing to TOM. For some reason the rebasing is giving me some build errors, so I will do things with a fresh clone. After that is done, this should be good to merge.

**Comment by ramiro050** [comment]
Are the python bindings for `torch` MLIR no longer available?

**Comment by silvasean** [comment]
> Are the python bindings for torch MLIR no longer available?

Not sure what you mean. Can you be more specific?

**Comment by ramiro050** [comment]
Never mind. I did not realize that if you don't run `ninja check-torch-mlir` after running `ninja`, the python bindings for `torch_mlir` are not built.

**Comment by ramiro050** [comment]
## PR Update

I've updated the imports to the new naming conventions in #311 . 

Everything now works as before, so this should be good to merge!

**Comment by silvasean** [comment]
one last thing before we merge this -- can you remove "iree" from the name?

Specifically, let's just call it `examples/torchfx/`

**Comment by ramiro050** [comment]
Done!
